api/client.go

```diff
@@ -329,13 +331,52 @@ func (c *Client) AddLocalCharm(curl *charm.URL, ch charm.Charm) (*charm.URL, err
 		return nil, errors.Errorf("unknown charm type %T", ch)
 	}
 
-	curl, err := c.UploadCharm(curl, archive)
+	anyHooks, err := hasHooks(archive.Name())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !anyHooks {
+		return nil, errors.Errorf("invalid charm %q: has no hooks", curl.Name)
+	}
+
+	curl, err = c.UploadCharm(curl, archive)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	return curl, nil
 }
 
+var hasHooks = hasHooksFolder
+
+func hasHooksFolder(name string) (bool, error) {
+	zipr, err := zip.OpenReader(name)
+	if err != nil {
+		return false, err
+	}
+	defer zipr.Close()
+
+	count := 0
+	// Cater for file separator differences between operating systems.
+	hooksPath := filepath.FromSlash("hooks/")
+	for _, f := range zipr.File {
+		if strings.Contains(f.Name, hooksPath) {
+			count++
```

Comment:
Why not just return true here? I don't get why we need count at all?

Reply:
I explained in the comment that finding just 1 occurrence of "hooks/" is not enough :) It could be just an empty folder called "hooks". We need to see more than one reference to "hooks/" in the file names.